tests/ad-tests.dx

```diff
@@ -322,3 +322,36 @@ func = \x:Float.
 
 snd (linearize func 1.0) 2.0
 > 2.
+
+-- Nested AD, examples from #713
+def func2 {m} (x:m=>Float) : Float =
+    exp (0.5 * sum for i. sq x.i)
+
+def hvpf {m} (x:m=>Float) (v:m=>Float) : m=>Float =
+    (dot x v) .* (grad func2 x) + func2(x) .* v
+
+x = [0.1, 0.2, 0.3]
+v = [0.2, 0.3, 0.4]
+func2 x
+> 1.072508
+
+-- analytic
+hvpf x v
+> [0.235952, 0.364653, 0.493354]
+
+-- finite diff over reverse
+eps = 0.0001
+(grad func2 (x + eps .* v) - grad func2 x) / eps
+> [0.235885, 0.364631, 0.493228]
+
+-- reverse over forward
+grad (\x. jvp func2 x v) x
+> [0.235952, 0.364653, 0.493354]
+
+-- reverse over reverse
+grad (\x. dot (grad func2 x) v) x
+> [0.235952, 0.364653, 0.493354]
+
+-- forward over reverse
+jvp (grad func2) x v
+> [0.235952, 0.364653, 0.493354]
```

Comment:
Instead of embedding those numbers in outputs, could you please write it as:
```
jvp (grad func2) x v ~~ [0.235952, 0.364653, 0.493354]
> True
```

The `~~` function computes approximate equality, which is what we should be using here.

Reply:
Done.